pkg/config/config.go

```diff
@@ -320,3 +361,46 @@ http_unix_socket:
 # What web browser gh should use when opening URLs. If blank, will refer to environment.
 browser:
 `
+
+//go:generate moq -rm -out migration_mock.go . Migration
+type Migration interface {
+	Do(*Config) (bool, error)
+}
+
+func Migrate(c *Config, m Migration) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Make a deep copy of the config so that migration can occur without being a destructive operation in the
+	// case of errors.
+	copiedCfg := c.deepCopyEntries()
+
+	requiredMigration, err := m.Do(copiedCfg)
+	if err != nil {
+		return fmt.Errorf("failed to migrate config: %s", err)
+	}
+
+	// If there was no migration required, we're done
+	if !requiredMigration {
+		return nil
+	}
+
+	// Otherwise we'll write our current hosts config to a backup file
+	if err := writeAssumingLockHeld(c, asBackup()); err != nil {
+		return fmt.Errorf("failed to write config backup after migration: %s", err)
+	}
+
+	// Then write out our migrated config. Note that since this is a deep copy, all the modified
+	// flags will be set to true, so we'll write out both config and hosts unconditionally.
+	if err := writeAssumingLockHeld(copiedCfg); err != nil {
+		// Note that this error case is not tested as it was not easily possible to force an error
+		// here, and not on the backup write above, without introducing new abstractions which
+		// seemed to complicate the whole affair.
+		return fmt.Errorf("failed to write config after migration: %s", err)
+	}
+
+	// Finally, swap in the copied config entries last, since we know the file has been successfully
+	// persisted.
+	c.entries = copiedCfg.entries
```

Comment:
I'm not sure if I understand you totally but I don't think we can return the migrated config because of the `sync.Once` package initialisation shenanigans. All places in the application point to the same config (unless someone has gone out of their way to avoid that).

Reply:
Hmm that is a good point around `sync.Once`, lets leave this as is then.